tests/unit/linalg/test_direct_solvers.cpp

```diff
@@ -138,19 +141,40 @@ TEST_CASE("Serial Direct Solvers", "[CUDA]")
       Vector B, X;
       a.FormLinearSystem(ess_tdof_list, x, b, A, X, B);
 
-      UMFPackSolver umf_solver;
-      umf_solver.Control[UMFPACK_ORDERING] = UMFPACK_ORDERING_METIS;
-      umf_solver.SetOperator(*A);
-      umf_solver.Mult(B, X);
+#ifdef MFEM_USE_SUITESPARSE
+      {
+         UMFPackSolver umf_solver;
+         umf_solver.Control[UMFPACK_ORDERING] = UMFPACK_ORDERING_METIS;
+         umf_solver.SetOperator(*A);
+         umf_solver.Mult(B, X);
+
+         Vector Y(X.Size());
+         A->Mult(X, Y);
+         Y -= B;
+         REQUIRE(Y.Norml2() < 1.e-12);
+
+         a.RecoverFEMSolution(X, b, x);
+         VectorFunctionCoefficient grad(dim, gradexact);
+         double error = x.ComputeH1Error(&uex, &grad);
+      }
+#endif
+#ifdef MFEM_USE_MKL_PARDISO
+      {
+         PardisoSolver pardiso_solver;
+         pardiso_solver.SetOperator(*A);
+         pardiso_solver.Mult(B, X);
+
+         Vector Y(X.Size());
+         A->Mult(X, Y);
+         Y -= B;
+         REQUIRE(Y.Norml2() < 1.e-12);
 
-      Vector Y(X.Size());
-      A->Mult(X, Y);
-      Y -= B;
-      REQUIRE(Y.Norml2() < 1.e-12);
+         a.RecoverFEMSolution(X, b, x);
+         VectorFunctionCoefficient grad(dim, gradexact);
+         double error = x.ComputeH1Error(&uex, &grad);
```

Comment:
This variable is goes out of scope on the following line, so there should be a compiler error on 154/178.  When I compile with cmake, I don't see the error because `MFEM_USE_MKL_PARDISO` is not defined, so `DIRECT_SOLVE_SERIAL` is also not defined.  When I build with SuiteSparse (so that `DIRECT_SOLVE_SERIAL` is defined), I get the expected compilation error.

@adam-sim-dev, do you use the makefile or cmake build system?

Reply:
To fix the cmake build system, you can apply the following patch
```git
diff --git a/CMakeLists.txt b/CMakeLists.txt
index e9e6ae0..7e30e5c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -473,6 +473,10 @@ if (MFEM_USE_MKL_CPARDISO)
   endif()
 endif()
 
+if (MFEM_USE_MKL_PARDISO)
+  find_package(MKL_PARDISO REQUIRED MKL_SEQUENTIAL MKL_LP64 MKL_MPI_WRAPPER)
+endif()
+
 # PARELAG
 if (MFEM_USE_PARELAG)
    find_package(PARELAG REQUIRED)
@@ -522,8 +526,8 @@ find_package(Threads REQUIRED)
 set(MFEM_TPLS OPENMP HYPRE LAPACK BLAS SuperLUDist STRUMPACK METIS SuiteSparse
     SUNDIALS PETSC SLEPC MUMPS AXOM FMS CONDUIT Ginkgo GNUTLS GSLIB
     NETCDF MPFR PUMI HIOP POSIXCLOCKS MFEMBacktrace ZLIB OCCA CEED RAJA UMPIRE
-    ADIOS2 CUBLAS CUSPARSE MKL_CPARDISO AMGX CALIPER CODIPACK BENCHMARK PARELAG
-    MPI_CXX HIP HIPSPARSE MOONOLITH BLITZ ALGOIM ENZYME)
+    ADIOS2 CUBLAS CUSPARSE MKL_CPARDISO MKL_PARDISO AMGX CALIPER CODIPACK
+    BENCHMARK PARELAG MPI_CXX HIP HIPSPARSE MOONOLITH BLITZ ALGOIM ENZYME)
 
 # Add all *_FOUND libraries in the variable TPL_LIBRARIES.
 set(TPL_LIBRARIES "")
```

To fix the unit test, you can apply the following patch
```bash
diff --git a/tests/unit/linalg/test_direct_solvers.cpp b/tests/unit/linalg/test_direct_solvers.cpp
index 2c40952..48a1ac3 100644
--- a/tests/unit/linalg/test_direct_solvers.cpp
+++ b/tests/unit/linalg/test_direct_solvers.cpp
@@ -156,6 +156,7 @@ TEST_CASE("Serial Direct Solvers", "[CUDA]")
          a.RecoverFEMSolution(X, b, x);
          VectorFunctionCoefficient grad(dim, gradexact);
          double error = x.ComputeH1Error(&uex, &grad);
+         REQUIRE(error < 1.e-12);
       }
 #endif
 #ifdef MFEM_USE_MKL_PARDISO
@@ -172,10 +173,9 @@ TEST_CASE("Serial Direct Solvers", "[CUDA]")
          a.RecoverFEMSolution(X, b, x);
          VectorFunctionCoefficient grad(dim, gradexact);
          double error = x.ComputeH1Error(&uex, &grad);
+         REQUIRE(error < 1.e-12);
       }
 #endif
-
-      REQUIRE(error < 1.e-12);
    }
 }
```